File: src/main/java/io/github/jhipster/sample/service/DiseaseMapService.java
package io.github.jhipster.sample.service;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import javax.persistence.EntityManager;
import javax.persistence.TypedQuery;
import javax.persistence.criteria.CriteriaBuilder;
import javax.persistence.criteria.CriteriaQuery;
import javax.persistence.criteria.Predicate;
import javax.persistence.criteria.Root;
import javax.transaction.Transactional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import io.github.jhipster.sample.domain.DiseaseBranch;
import io.github.jhipster.sample.domain.DiseaseMap;
import io.github.jhipster.sample.domain.DiseaseMapIndexDTO;
import io.github.jhipster.sample.domain.DiseaseXiAn;
import io.github.jhipster.sample.domain.LinkCard;
import io.github.jhipster.sample.domain.QArobot;
import io.github.jhipster.sample.repository.DiseaseBranchRepository;
import io.github.jhipster.sample.repository.DiseaseMapRepository;
import io.github.jhipster.sample.repository.DiseaseXiAnRepository;
import io.github.jhipster.sample.repository.LinkCardRepository;
import io.github.jhipster.sample.repository.QArobotRepository;
import io.github.jhipster.sample.web.rest.searchdto.DiseaseMapSearchDTO;
import io.github.jhipster.sample.web.rest.util.SearchUtil;

/**
 * DiseaseMapService used to associate disease map with diseasexian and qarobot.
 * It can also used to create new disease map and associate the newly created
 * disease map with the parent disease map.
 *
 * Another job is to mount new disease branch and delete disease branch. attach
 * new disease map to disease branch.
 */
@Service
public class DiseaseMapService {
    private final DiseaseXiAnRepository diseaseXiAnRepository;
    private final QArobotRepository qArobotRepository;
    private final DiseaseMapRepository diseaseMapRepository;
    private final DiseaseBranchRepository diseaseBranchRepository;
    private final LinkCardRepository linkCardRepository;
    private final EntityManager entityManager;
    @Autowired
    public DiseaseMapService(
        DiseaseMapRepository diseaseMapRepository,
        DiseaseBranchRepository diseaseBranchRepository,
        DiseaseXiAnRepository diseaseXiAnRepository,
        QArobotRepository qArobotRepository,
        LinkCardRepository linkCardRepository,
        EntityManager entityManager) {
        this.diseaseXiAnRepository = diseaseXiAnRepository;
        this.diseaseMapRepository = diseaseMapRepository;
        this.qArobotRepository = qArobotRepository;
        this.diseaseBranchRepository = diseaseBranchRepository;
        this.linkCardRepository = linkCardRepository;
        this.entityManager = entityManager;
    }

    /**
     * Attach new disease branch.
     *
     * @param newDiseaseBranch
     * @param diseaseBranchId
     */
    @Transactional
    public DiseaseBranch attachDiseaseBranch(DiseaseBranch newDiseaseBranch) {
        return diseaseBranchRepository.save(newDiseaseBranch);
    }

    /**
     * get All disease branch.
     *
     * @return
     */
    public Page<DiseaseBranch> getAllDiseaseBranchPageable(Pageable pageable) {
        Page<DiseaseBranch> page = diseaseBranchRepository.findAll(pageable);
        return page;
    }

    /**
     * get All disease branch.
     *
     * @return
     */
    public List<DiseaseBranch> getAllDiseaseBranch() {
        List<DiseaseBranch> page = diseaseBranchRepository.findAll();
        return page;
    }

    /**
     * get disease branch.
     *
     * @return
     */
    public DiseaseBranch getDiseaseBranch(Long diseaseBranchId) {
        return diseaseBranchRepository.findById(diseaseBranchId).get();
    }

    /**
     * get disease branch.
     *
     * @return
     */
    @Transactional
    public DiseaseBranch getDiseaseBranchEagerly(Long diseaseBranchId) {
        DiseaseBranch diseaseBranch = diseaseBranchRepository.findById(diseaseBranchId).get();
        diseaseBranch.getDiseaseMaps().size();
        for (DiseaseMap map : diseaseBranch.getDiseaseMaps()) {
            map.getDiseaseMaps().size();
        }
        return diseaseBranch;
    }

    /**
     * get child disease maps of disease map.
     *
     * @return
     */
    @Transactional
    public List<DiseaseMap> getDiseaseMaps(Long diseaseMapId) {
        DiseaseMap diseaseMap = diseaseMapRepository.findById(diseaseMapId).get();
        diseaseMap.getDiseaseMaps().size();
        return diseaseMap.getDiseaseMaps();
    }

    /**
     * get disease map.
     *
     * @return
     */
    // @Transactional
    public DiseaseMap getDiseaseMap(Long diseaseMapId) {
        DiseaseMap map = diseaseMapRepository.findById(diseaseMapId).get();
        map.setDiseaseXiAns(null);
        map.setQarobots(null);
        return map;

    }

    @Transactional
    public DiseaseMap getDiseaseMapEagerly(Long diseaseMapId) {
        DiseaseMap map = diseaseMapRepository.findById(diseaseMapId).get();
        map.getDiseaseXiAns().size();
        map.getQarobots().size();
        DiseaseMap dto = new DiseaseMap();
        dto.setDiseaseXiAns(map.getDiseaseXiAns());
        dto.setQarobots(map.getQarobots());
        return dto;
    }

    /**
     * get all disease maps in a disease branch
     *
     * @param diseaseBranchId
     * @return
     */
    @org.springframework.transaction.annotation.Transactional
    public List<DiseaseMap> getAllDiseaseMap(Long diseaseBranchId) {

        List<DiseaseMap> diseaseMaps = diseaseBranchRepository.findOneWithDiseaseMapsById(diseaseBranchId).get()
                .getDiseaseMaps();

        return diseaseMaps;
    }

    /**
     * Deattach disease branch
     *
     * @param diseaseBranchId
     */
    @Transactional
    public void deattachDiseaseBranch(Long diseaseBranchId) {
        diseaseBranchRepository.deleteById(diseaseBranchId);
    }

    /**
     * Add new disease map to disease branch specified with diseaseBranchId.
     *
     * @param diseaseMap
     * @param diseaseBranchId
     */
    @Transactional
    public void attachDiseaseMapToDiseaseBranch(DiseaseMap diseaseMap, Long diseaseBranchId) {
        DiseaseMap map = diseaseMapRepository.save(diseaseMap);
        diseaseBranchRepository.findById(diseaseBranchId).get().getDiseaseMaps().add(map);
    }

    /**
     * Put Disease Map what already exists in database to modify some content.
     *
     * @param diseaseMap
     */
    @Transactional
    public void modifyDiseaseMap(DiseaseMap diseaseMap) {
        modifyDiseaseMap(diseaseMapRepository.findById(diseaseMap.getId()).get(), diseaseMap);
    }

    /**
     * Modify disease Branch
     *
     * @param diseaseMap
     */
    @Transactional
    public void modifyDiseaseBranch(DiseaseBranch diseaseBranch) {
        modifyDiseaseBranch(diseaseBranchRepository.findById(diseaseBranch.getId()).get(),
                diseaseBranch);
    }

    /**
     * delete a disease map. all sub disease map related to this map would be
     * deleted.
     *
     * @param diseaseMapId
     */
    @Transactional
    public void deleteDiseaseMap(Long diseaseMapId) {
        diseaseMapRepository.deleteById(diseaseMapId);
    }

    /**
     * associate diseaseMap with a diseaseXiAn
     *
     * @param diseaseMapId
     * @param diseaseXiAnId
     */
    @Transactional
    public void associateWithDiseaseXiAn(Long diseaseMapId, Long diseaseXiAnId) {
        DiseaseMap diseaseMap = diseaseMapRepository.findById(diseaseMapId).get();
        DiseaseXiAn diseaseXiAn = diseaseXiAnRepository.findById(diseaseXiAnId).get();
        diseaseMap.getDiseaseXiAns().add(diseaseXiAn);
    }

    /**
     * deassociate diseaseMap with a diseaseXiAn
     *
     * @param diseaseMapId
     * @param diseaseXiAnId
     */
    @Transactional
    public void deassociateWithDiseaseXiAn(Long diseaseMapId, Long diseaseXiAnId) {
        DiseaseMap diseaseMap = diseaseMapRepository.findById(diseaseMapId).get();
        DiseaseXiAn diseaseXiAn = diseaseXiAnRepository.findById(diseaseXiAnId).get();
        diseaseMap.getDiseaseXiAns().remove(diseaseXiAn);
    }

    /**
     * associate diseaseMap with QArobot.
     *
     * @param diseaseMapId
     * @param qArobotId
     */
    @Transactional
    public void associatedWithQArobot(Long diseaseMapId, Long qArobotId) {
        DiseaseMap diseaseMap = diseaseMapRepository.findById(diseaseMapId).get();
        QArobot qArobot = qArobotRepository.findById(qArobotId).get();
        diseaseMap.getQarobots().add(qArobot);
    }

    /**
     * deassociate diseaseMap with QArobot.
     *
     * @param diseaseMapId
     * @param qArobotId
     */
    @Transactional
    public void deassociatedWithQArobot(Long diseaseMapId, Long qArobotId) {
        DiseaseMap diseaseMap = diseaseMapRepository.findById(diseaseMapId).get();
        QArobot qArobot = qArobotRepository.findById(qArobotId).get();
        diseaseMap.getQarobots().remove(qArobot);
    }

    /**
     * attach disease map to a disease map.
     *
     * @param newDiseaseMap
     * @param diseaseMapId
     */
    @Transactional
    public void attachDiseaseMapToDiseaseMap(DiseaseMap newDiseaseMap, Long diseaseMapId) {
        DiseaseMap newMap = diseaseMapRepository.save(newDiseaseMap);
        diseaseMapRepository.findById(diseaseMapId).get().getDiseaseMaps().add(newMap);
    }

    /**
     * attach link card to a disease map.
     *
     * @param linkcard
     * @param diseaseMapId
     */
    public void attachLinkCardToDiseaseMap(LinkCard linkcard, Long diseaseMapId) {
        DiseaseMap diseaseMap = diseaseMapRepository.findById(diseaseMapId).get();
        Set<LinkCard> links = diseaseMap.getLinkCards();
        if (links.contains(linkcard)) {
            linkCardRepository.save(linkcard);
        } else {
            diseaseMap.getLinkCards().add(linkcard);
            diseaseMapRepository.save(diseaseMap);
        }
    }

    /**
     * attach link card to a disease map.
     *
     * @param linkcard
     * @param diseaseMapId
     */
    @Transactional
    public void deattachLinkCardToDiseaseMap(LinkCard linkcard, Long diseaseMapId) {
        diseaseMapRepository.findById(diseaseMapId).get().getLinkCards().remove(linkcard);
    }

    /**
     * attach link card to a disease branch
     *
     * @param linkcard
     * @param diseaseBranchId
     */
    public void attachLinkCardToDiseaseBranch(LinkCard linkcard, Long diseaseBranchId) {
        DiseaseBranch diseaseBranch = diseaseBranchRepository.findById(diseaseBranchId).get();
        Set<LinkCard> links = diseaseBranch.getLinkCards();
        if (links.contains(linkcard)) {
            linkCardRepository.save(linkcard);
        } else {
            diseaseBranch.getLinkCards().add(linkcard);
            diseaseBranchRepository.save(diseaseBranch);
        }
    }

    /**
     * deattach link card to a disease branch
     *
     * @param linkcard
     * @param diseaseBranchId
     */
    @Transactional
    public void deattachLinkCardToDiseaseBranch(LinkCard linkcard, Long diseaseBranchId) {
        diseaseBranchRepository.findById(diseaseBranchId).get().getLinkCards().remove(linkcard);
    }

    @Transactional
    public Page<DiseaseBranch> searchDiseaseBranch(DiseaseMapSearchDTO searchDTO, Pageable pageable) {
        CriteriaBuilder cb = entityManager.getCriteriaBuilder();
        CriteriaQuery<DiseaseBranch> qarobotQuery = cb.createQuery(DiseaseBranch.class);
        CriteriaQuery<Long> countQuery = cb.createQuery(Long.class);
        Root<DiseaseBranch> qarobot = qarobotQuery.from(DiseaseBranch.class);
        Root<DiseaseBranch> count = countQuery.from(DiseaseBranch.class);

        List<Predicate> restrictions = new ArrayList<Predicate>();
        qarobotQuery.select(qarobot);
        countQuery.select(cb.count(count));

        String query = searchDTO.getName();
        if (query != null && query.length() > 0) {
            restrictions.clear();
            restrictions.addAll(SearchUtil.queryKeywordParser(query).stream().map(
                keyword -> cb.like(qarobot.get("name"), "%" + keyword + "%")
                ).collect(Collectors.toList()));
            restrictions.add(cb.like(qarobot.get("name"), "%" + query + "%"));
        }

        Predicate queryPredicate =
            restrictions.size() > 0 ?
                cb.or(restrictions.toArray(new Predicate[restrictions.size()])) : cb.and();

        qarobotQuery.where(queryPredicate);

        // get qarobots satisfied with criterias
        TypedQuery<DiseaseBranch> typedDiseaseQuery = entityManager.createQuery(qarobotQuery);
        // typedDiseaseQuery.setFirstResult((int)pageable.getOffset());
        // typedDiseaseQuery.setMaxResults((int)pageable.getPageSize());
        List<DiseaseBranch> allDis = typedDiseaseQuery.getResultList();

        // get totalItems number with criterias
        TypedQuery<Long> typedCountQuery = entityManager.createQuery(countQuery);
        Long totalItems = typedCountQuery.getSingleResult();

        // create a page in terms of the count and content
        Page<DiseaseBranch> resultPage = new PageImpl<>(allDis, pageable, totalItems);

        return resultPage;
    }

    @Transactional
    public Page<DiseaseMapIndexDTO> searchDiseaseMap(DiseaseMapSearchDTO searchDTO, Pageable pageable) {
        CriteriaBuilder cb = entityManager.getCriteriaBuilder();
        CriteriaQuery<DiseaseMap> qarobotQuery = cb.createQuery(DiseaseMap.class);
        CriteriaQuery<Long> countQuery = cb.createQuery(Long.class);
        Root<DiseaseMap> qarobot = qarobotQuery.from(DiseaseMap.class);
        Root<DiseaseMap> count = countQuery.from(DiseaseMap.class);

        List<Predicate> restrictions = new ArrayList<Predicate>();
        qarobotQuery.select(qarobot);
        countQuery.select(cb.count(count));

        String query = searchDTO.getName();
        if (query != null && query.length() > 0) {
            restrictions.clear();
            restrictions.addAll(SearchUtil.queryKeywordParser(query).stream().map(
                keyword -> cb.like(qarobot.get("name"), "%" + keyword + "%")
                ).collect(Collectors.toList()));
            restrictions.add(cb.like(qarobot.get("name"), "%" + query + "%"));
        }

        Predicate queryPredicate =
            restrictions.size() > 0 ?
                cb.or(restrictions.toArray(new Predicate[restrictions.size()])) : cb.and();

        qarobotQuery.where(queryPredicate);

        // get qarobots satisfied with criterias
        TypedQuery<DiseaseMap> typedDiseaseQuery = entityManager.createQuery(qarobotQuery);
        typedDiseaseQuery.setFirstResult((int)pageable.getOffset());
        typedDiseaseQuery.setMaxResults((int)pageable.getPageSize());
        List<DiseaseMap> allDis = typedDiseaseQuery.getResultList();

        // get totalItems number with criterias
        TypedQuery<Long> typedCountQuery = entityManager.createQuery(countQuery);
        Long totalItems = typedCountQuery.getSingleResult();

        // create a page in terms of the count and content
        Page<DiseaseMapIndexDTO> resultPage = new PageImpl<DiseaseMapIndexDTO>(allDis.stream().map(any -> any.toIndexDTO()).collect(Collectors.toList()), pageable, totalItems);

        return resultPage;
    }

    private DiseaseMap modifyDiseaseMap(DiseaseMap oldMap, DiseaseMap newMap) {
        oldMap.setDescription(newMap.getDescription());
        oldMap.setId(newMap.getId());
        oldMap.setName(newMap.getName());
        oldMap.setSubsidiary(newMap.getSubsidiary());
        return oldMap;
    }

    private DiseaseBranch modifyDiseaseBranch(DiseaseBranch oldBranch, DiseaseBranch newBranch) {
        oldBranch.setDescription(newBranch.getDescription());
        oldBranch.setId(newBranch.getId());
        oldBranch.setName(newBranch.getName());
        oldBranch.setSubsidiary(newBranch.getSubsidiary());
        return oldBranch;
    }

    public DiseaseMapIndexDTO diseaseMapIndexConverter(DiseaseMap diseaseMap) {
        DiseaseMapIndexDTO dto = new DiseaseMapIndexDTO();
        dto.setId(diseaseMap.getId());
        dto.setName(diseaseMap.getName());
        dto.setDescription(diseaseMap.getDescription());
        dto.setSubsidiary(diseaseMap.getSubsidiary());
        return dto;
    }

}

File: src/main/webapp/app/disease-map/disease-map.service.ts
import { ILinkCard } from './../shared/model/link-card.model';
import { IDiseaseBranch, DiseaseBranch } from './../shared/model/disease-branch.model';
import { HttpClient, HttpResponse, HttpParams, HttpEventType } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { SERVER_API_URL } from 'app/app.constants';
import { Observable } from 'rxjs';
import { IFile } from 'app/shared/model/file.model';
import { map } from 'rxjs/operators';
import { DiseaseMap, IDiseaseMap } from 'app/shared/model/disease-map.model';
import { createRequestOption } from 'app/shared';
import { DiseasePartition } from 'app/shared/model/disease-partition.model';

type BranchResponseType = HttpResponse<IDiseaseBranch>;
type BranchResponseArrayType = HttpResponse<IDiseaseBranch[]>;
type MapResponseType = HttpResponse<IDiseaseMap>;
type MapResponseArrayType = HttpResponse<IDiseaseMap[]>;

@Injectable({
    providedIn: 'root'
})
export class DiseaseMapService {

    // url for suppliess
    public diseaseMapUrl = SERVER_API_URL + 'api/disease-map';

    public diseasePartitonUrl = SERVER_API_URL + 'api/disease-partitions';

    constructor(protected http: HttpClient) { }

    attachDiseaseBranch(diseaseBranch: IDiseaseBranch) {
        return this.http.post<any>(`${this.diseaseMapUrl}/attach-disease-branch`, diseaseBranch, { observe: 'body' });
    }

    getAllDiseaseBranch(req: any): Observable<BranchResponseArrayType> {
        const _params = createRequestOption(req);
        return this.http.get<IDiseaseBranch[]>(`${this.diseaseMapUrl}/get-all-disease-branch`, { params:  _params, observe: 'response'});
    }

    getDiseaseBranch(diseaseBranchId: number): Observable<BranchResponseType> {
        return this.http.get<IDiseaseBranch>(`${this.diseaseMapUrl}/get-disease-branch/${diseaseBranchId}`, { observe: 'response'});
    }

    getDiseaseBranchEagerly(diseaseBranchId: number): Observable<BranchResponseType> {
        return this.http.get<IDiseaseBranch>(`${this.diseaseMapUrl}/get-disease-branch-eagerly/${diseaseBranchId}`, { observe: 'response'});
    }

    getDiseaseMap(diseaseMapId: number): Observable<MapResponseType> {
        return this.http.get<IDiseaseMap>(`${this.diseaseMapUrl}/get-disease-map/${diseaseMapId}`, { observe: 'response'});
    }

    /**
     * Get disease map eagerly with diseases and qarobots
     * @param diseaseMapId
     */
    getDiseaseMapEagerly(diseaseMapId: number): Observable<MapResponseType> {
        return this.http.get<IDiseaseMap>(`${this.diseaseMapUrl}/get-disease-map-eagerly/${diseaseMapId}`, { observe: 'response'});
    }

    /**
     * get child disease maps of disease mp.
     * @param diseaseMapId diseaseMapId
     */
    getDiseaseMaps(diseaseMapId: number): Observable<MapResponseArrayType> {
        return this.http.get<IDiseaseMap[]>(`${this.diseaseMapUrl}/get-disease-maps/${diseaseMapId}`, { observe: 'response' });
    }

    /**
     * Get all children disease map of diseae branch.
     * @param diseaseBranchId
     */
    getAllDiseaseMap(diseaseBranchId: number) {
        return this.http.get<IDiseaseMap[]>(`${this.diseaseMapUrl}/get-all-disease-map/${diseaseBranchId}`, { observe: 'body'});
    }

    addLinkToMap(link: ILinkCard, _map: IDiseaseMap) {
        return this.http.post<any>(`${this.diseaseMapUrl}/attach-link-card-to-disease-map/${_map.id}`, link);
    }

    deleteLinkToMap(link: ILinkCard, _map: IDiseaseMap) {
        return this.http.post<any>(`${this.diseaseMapUrl}/deattach-link-card-to-disease-map/${_map.id}`, link);
    }

    addLinkToBranch(link: ILinkCard, branch: IDiseaseBranch) {
        return this.http.post<any>(`${this.diseaseMapUrl}/attach-link-card-to-disease-branch/${branch.id}`, link);
    }

    deleteLinkToBranch(link: ILinkCard, branch: IDiseaseBranch) {
        return this.http.post<any>(`${this.diseaseMapUrl}/deattach-link-card-to-disease-branch/${branch.id}`, link);
    }

    deattachDiseaseBranch(diseaseBranchId: number) {
        return this.http.delete<IDiseaseMap[]>(`${this.diseaseMapUrl}/deattach-disease-branch/${diseaseBranchId}`, { observe: 'body'});
    }

    attachDiseaseMapToDiseaseBranch(diseaseMap: DiseaseMap, diseaseBranchId: number) {
        return this.http.post<any>(`${this.diseaseMapUrl}/attach-disease-map-to-disease-branch/${diseaseBranchId}`, diseaseMap, {observe: 'body'});
    }

    attachDiseaseMapToDiseaseMap(diseaseMap: DiseaseMap, diseaseMapId: number) {
        return this.http.post<any>(`${this.diseaseMapUrl}/attach-disease-map-to-disease-map/${diseaseMapId}`, diseaseMap, {observe: 'body'});
    }

    associateWithQArobot(diseaseMapId: number, qArobotId: number) {
        return this.http.get<any>(`${this.diseaseMapUrl}/associate-with-q-arobot/${diseaseMapId}/${qArobotId}`, {observe: 'response'});
    }

    associateWithDiseaseXiAn(diseaseMapId: number, diseaseXiAnId: number) {
        return this.http.get<any>(`${this.diseaseMapUrl}/assocaite-with-disease-xi-an/${diseaseMapId}/${diseaseXiAnId}`, {observe: 'response'});
    }

    deassociateWithQArobot(diseaseMapId: number, qArobotId: number) {
        return this.http.get<any>(`${this.diseaseMapUrl}/deassociate-with-q-arobot/${diseaseMapId}/${qArobotId}`, {observe: 'response'});
    }

    deassociateWithDiseaseXiAn(diseaseMapId: number, diseaseXiAnId: number) {
        return this.http.get<any>(`${this.diseaseMapUrl}/deassocaite-with-disease-xi-an/${diseaseMapId}/${diseaseXiAnId}`, {observe: 'response'});
    }

    deleteDiseaseMap(diseaseMapId: number) {
        return this.http.delete<any>(`${this.diseaseMapUrl}/delete-disease-map/${diseaseMapId}`, {observe: 'response'});
    }

    modifyDiseaseMap(diseaseMap: IDiseaseMap) {
        return this.http.put<any>(`${this.diseaseMapUrl}/modify-disease-map`, diseaseMap, { observe: 'response'});
    }

    modifyDiseaseBranch(diseaseBranch: IDiseaseBranch) {
        return this.http.put<any>(`${this.diseaseMapUrl}/modify-disease-branch`, diseaseBranch, {observe: 'response'});
    }

    searchDiseaseBranch(req: any): Observable<BranchResponseArrayType> {
        const _params = createRequestOption(req);
        return this.http.get<IDiseaseBranch[]>(`${this.diseaseMapUrl}/_search-disease-branch`, { params: _params, observe: 'response' });
    }

    searchDiseaseMap(req: any): Observable<MapResponseArrayType> {
        const _params = createRequestOption(req);
        return this.http.get<IDiseaseMap[]>(`${this.diseaseMapUrl}/_search-disease-map`, { params: _params, observe: 'response' });
    }

    // Url for partitions:
    getDiseasePartitions(): Observable<DiseasePartition[]> {
        return this.http.get<DiseasePartition[]>(`${this.diseasePartitonUrl}`, { observe: 'body' });
    }

    getDiseaePartition(partitionId: number): Observable<DiseasePartition> {
        return this.http.get<DiseasePartition>(`${this.diseasePartitonUrl}/${partitionId}`, { observe: 'body' });
    }

    postDiseasePartition(diseasePartition: DiseasePartition): Observable<void> {
        return this.http.post<void>(`${this.diseasePartitonUrl}`, { body: diseasePartition });
    }

    postDiseasePartitionsDiseaseBranch(diseasePartitionId: number, diseaseBranch: DiseaseBranch): Observable<void> {
        return this.http.post<void>(`${this.diseasePartitonUrl}/${diseasePartitionId}/disease-branches`, { body: diseaseBranch });
    }

}
